Refactor regexp costructor

raskad opened the pull request:
This Pull Request relates to #1304.

It changes the following:

- Refactor the regexp constructor to match the spec
- Fix various errors in regexp functions

The test `built-ins/RegExp/prototype/Symbol.split/limit-0-bail.js` is now failing, because `Regexp[Symbol.split]` gets the `flags` property from the regexp prototype, which is undefined. I'm not sure where the root cause for that is, as everything in the immediate call path seems spec compliant to me. But I think we can fix that later as the split implementation itself seems spec compliant.

raskad commented:
| Test result | master count | PR count | difference |
  | :---------: | :----------: | :------: | :--------: |
  | Total | 78,897 | 78,897 | 0 |
  | Passed | 28,432 | 28,616 | **+184** |
  | Ignored | 15,614 | 15,614 | 0 |
  | Failed | 34,851 | 34,667 | **-184** |
  | Panics | 2 | 2 | 0 |
  | Conformance | 36.04% | 36.27% | **+0.23%** |
  
  <details><summary><b>Fixed tests:</b></summary>
  
  ```
  test/built-ins/Array/isArray/15.4.3.2-1-10.js [strict mode] (previously Failed)
  test/built-ins/Array/isArray/15.4.3.2-1-10.js (previously Failed)
  test/built-ins/Array/prototype/lastIndexOf/15.4.4.15-5-21.js [strict mode] (previously Failed)
  test/built-ins/Array/prototype/lastIndexOf/15.4.4.15-5-21.js (previously Failed)
  test/built-ins/Array/prototype/indexOf/15.4.4.14-1-12.js [strict mode] (previously Failed)
  test/built-ins/Array/prototype/indexOf/15.4.4.14-1-12.js (previously Failed)
  test/built-ins/Array/prototype/filter/15.4.4.20-1-12.js [strict mode] (previously Failed)
  test/built-ins/Array/prototype/filter/15.4.4.20-1-12.js (previously Failed)
  test/built-ins/Array/prototype/filter/15.4.4.20-9-c-iii-23.js [strict mode] (previously Failed)
  test/built-ins/Array/prototype/filter/15.4.4.20-9-c-iii-23.js (previously Failed)
  test/built-ins/Array/prototype/filter/15.4.4.20-5-16.js [strict mode] (previously Failed)
  test/built-ins/Array/prototype/filter/15.4.4.20-5-16.js (previously Failed)
  test/built-ins/Array/prototype/every/15.4.4.16-5-16.js [strict mode] (previously Failed)
  test/built-ins/Array/prototype/every/15.4.4.16-5-16.js (previously Failed)
  test/built-ins/Array/prototype/every/15.4.4.16-7-c-iii-22.js [strict mode] (previously Failed)
  test/built-ins/Array/prototype/every/15.4.4.16-7-c-iii-22.js (previously Failed)
  test/built-ins/Array/prototype/every/15.4.4.16-1-12.js [strict mode] (previously Failed)
  test/built-ins/Array/prototype/every/15.4.4.16-1-12.js (previously Failed)
  test/built-ins/Array/prototype/some/15.4.4.17-5-16.js [strict mode] (previously Failed)
  test/built-ins/Array/prototype/some/15.4.4.17-5-16.js (previously Failed)
  test/built-ins/Array/prototype/some/15.4.4.17-1-12.js [strict mode] (previously Failed)
  test/built-ins/Array/prototype/some/15.4.4.17-1-12.js (previously Failed)
  test/built-ins/Array/prototype/some/15.4.4.17-7-c-iii-22.js [strict mode] (previously Failed)
  test/built-ins/Array/prototype/some/15.4.4.17-7-c-iii-22.js (previously Failed)
  test/built-ins/Array/prototype/forEach/15.4.4.18-1-12.js [strict mode] (previously Failed)
  test/built-ins/Array/prototype/forEach/15.4.4.18-1-12.js (previously Failed)
  test/built-ins/Array/prototype/forEach/15.4.4.18-5-16.js [strict mode] (previously Failed)
  test/built-ins/Array/prototype/forEach/15.4.4.18-5-16.js (previously Failed)
  test/built-ins/Array/prototype/reduceRight/15.4.4.22-1-12.js [strict mode] (previously Failed)
  test/built-ins/Array/prototype/reduceRight/15.4.4.22-1-12.js (previously Failed)
  test/built-ins/Array/prototype/reduceRight/15.4.4.22-9-c-ii-32.js [strict mode] (previously Failed)
  test/built-ins/Array/prototype/reduceRight/15.4.4.22-9-c-ii-32.js (previously Failed)
  test/built-ins/Array/prototype/reduce/15.4.4.21-1-12.js [strict mode] (previously Failed)
  test/built-ins/Array/prototype/reduce/15.4.4.21-1-12.js (previously Failed)
  test/built-ins/Array/prototype/reduce/15.4.4.21-9-c-ii-32.js [strict mode] (previously Failed)
  test/built-ins/Array/prototype/reduce/15.4.4.21-9-c-ii-32.js (previously Failed)
  test/built-ins/Array/prototype/map/15.4.4.19-1-12.js [strict mode] (previously Failed)
  test/built-ins/Array/prototype/map/15.4.4.19-1-12.js (previously Failed)
  test/built-ins/Array/prototype/map/15.4.4.19-5-16.js [strict mode] (previously Failed)
  test/built-ins/Array/prototype/map/15.4.4.19-5-16.js (previously Failed)
  test/built-ins/Object/defineProperties/15.2.3.7-5-b-79.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperties/15.2.3.7-5-b-79.js (previously Failed)
  test/built-ins/Object/defineProperties/15.2.3.7-5-b-104.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperties/15.2.3.7-5-b-104.js (previously Failed)
  test/built-ins/Object/defineProperties/15.2.3.7-5-b-158.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperties/15.2.3.7-5-b-158.js (previously Failed)
  test/built-ins/Object/defineProperties/15.2.3.7-5-b-26.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperties/15.2.3.7-5-b-26.js (previously Failed)
  test/built-ins/Object/defineProperties/15.2.3.7-5-b-183.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperties/15.2.3.7-5-b-183.js (previously Failed)
  test/built-ins/Object/defineProperties/15.2.3.7-5-b-132.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperties/15.2.3.7-5-b-132.js (previously Failed)
  test/built-ins/Object/defineProperties/15.2.3.7-5-b-211.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperties/15.2.3.7-5-b-211.js (previously Failed)
  test/built-ins/Object/defineProperties/15.2.3.7-5-a-14.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperties/15.2.3.7-5-a-14.js (previously Failed)
  test/built-ins/Object/defineProperties/15.2.3.7-5-b-246.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperties/15.2.3.7-5-b-246.js (previously Failed)
  test/built-ins/Object/defineProperties/15.2.3.7-2-13.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperties/15.2.3.7-2-13.js (previously Failed)
  test/built-ins/Object/defineProperties/15.2.3.7-5-b-51.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperties/15.2.3.7-5-b-51.js (previously Failed)
  test/built-ins/Object/getPrototypeOf/15.2.3.2-2-26.js [strict mode] (previously Failed)
  test/built-ins/Object/getPrototypeOf/15.2.3.2-2-26.js (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-40.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-40.js (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-4-585.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-4-585.js (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-4-40.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-4-40.js (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-118.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-118.js (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-4-393.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-4-393.js (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-172-1.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-172-1.js (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-255.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-255.js (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-146.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-146.js (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-4-409.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-4-409.js (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-172.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-172.js (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-65.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-65.js (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-93.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-93.js (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-255-1.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-255-1.js (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-40-1.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-40-1.js (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-93-1.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-93-1.js (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-146-1.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-146-1.js (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-197.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-197.js (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-225.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-225.js (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-225-1.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-225-1.js (previously Failed)
  test/built-ins/Object/create/15.2.3.5-4-12.js [strict mode] (previously Failed)
  test/built-ins/Object/create/15.2.3.5-4-12.js (previously Failed)
  test/built-ins/Object/create/15.2.3.5-4-66.js [strict mode] (previously Failed)
  test/built-ins/Object/create/15.2.3.5-4-66.js (previously Failed)
  test/built-ins/Object/create/15.2.3.5-4-144.js [strict mode] (previously Failed)
  test/built-ins/Object/create/15.2.3.5-4-144.js (previously Failed)
  test/built-ins/Object/create/15.2.3.5-4-198.js [strict mode] (previously Failed)
  test/built-ins/Object/create/15.2.3.5-4-198.js (previously Failed)
  test/built-ins/Object/create/15.2.3.5-4-223.js [strict mode] (previously Failed)
  test/built-ins/Object/create/15.2.3.5-4-223.js (previously Failed)
  test/built-ins/Object/create/15.2.3.5-4-119.js [strict mode] (previously Failed)
  test/built-ins/Object/create/15.2.3.5-4-119.js (previously Failed)
  test/built-ins/Object/create/15.2.3.5-4-172.js [strict mode] (previously Failed)
  test/built-ins/Object/create/15.2.3.5-4-172.js (previously Failed)
  test/built-ins/Object/create/15.2.3.5-4-91.js [strict mode] (previously Failed)
  test/built-ins/Object/create/15.2.3.5-4-91.js (previously Failed)
  test/built-ins/Object/create/15.2.3.5-4-250.js [strict mode] (previously Failed)
  test/built-ins/Object/create/15.2.3.5-4-250.js (previously Failed)
  test/built-ins/Object/create/15.2.3.5-4-35.js [strict mode] (previously Failed)
  test/built-ins/Object/create/15.2.3.5-4-35.js (previously Failed)
  test/built-ins/Object/create/15.2.3.5-4-286.js [strict mode] (previously Failed)
  test/built-ins/Object/create/15.2.3.5-4-286.js (previously Failed)
  test/built-ins/Object/prototype/toString/Object.prototype.toString.call-regexp.js [strict mode] (previously Failed)
  test/built-ins/Object/prototype/toString/Object.prototype.toString.call-regexp.js (previously Failed)
  test/built-ins/String/prototype/matchAll/regexp-matchAll-is-undefined-or-null.js [strict mode] (previously Failed)
  test/built-ins/String/prototype/matchAll/regexp-matchAll-is-undefined-or-null.js (previously Failed)
  test/built-ins/String/prototype/matchAll/regexp-is-undefined-or-null-invokes-matchAll.js [strict mode] (previously Failed)
  test/built-ins/String/prototype/matchAll/regexp-is-undefined-or-null-invokes-matchAll.js (previously Failed)
  test/built-ins/String/prototype/matchAll/regexp-prototype-has-no-matchAll.js [strict mode] (previously Failed)
  test/built-ins/String/prototype/matchAll/regexp-prototype-has-no-matchAll.js (previously Failed)
  test/built-ins/String/prototype/match/S15.5.4.10_A1_T4.js [strict mode] (previously Failed)
  test/built-ins/String/prototype/match/S15.5.4.10_A1_T4.js (previously Failed)
  test/built-ins/String/prototype/match/invoke-builtin-match.js [strict mode] (previously Failed)
  test/built-ins/String/prototype/match/invoke-builtin-match.js (previously Failed)
  test/built-ins/RegExp/S15.10.4.1_A8_T13.js [strict mode] (previously Failed)
  test/built-ins/RegExp/S15.10.4.1_A8_T13.js (previously Failed)
  test/built-ins/RegExp/S15.10.4.1_A5_T7.js [strict mode] (previously Failed)
  test/built-ins/RegExp/S15.10.4.1_A5_T7.js (previously Failed)
  test/built-ins/RegExp/S15.10.4.1_A5_T1.js [strict mode] (previously Failed)
  test/built-ins/RegExp/S15.10.4.1_A5_T1.js (previously Failed)
  test/built-ins/RegExp/S15.10.4.1_A5_T9.js [strict mode] (previously Failed)
  test/built-ins/RegExp/S15.10.4.1_A5_T9.js (previously Failed)
  test/built-ins/RegExp/S15.10.3.1_A2_T2.js [strict mode] (previously Failed)
  test/built-ins/RegExp/S15.10.3.1_A2_T2.js (previously Failed)
  test/built-ins/RegExp/S15.10.3.1_A2_T1.js [strict mode] (previously Failed)
  test/built-ins/RegExp/S15.10.3.1_A2_T1.js (previously Failed)
  test/built-ins/RegExp/S15.10.4.1_A7_T2.js [strict mode] (previously Failed)
  test/built-ins/RegExp/S15.10.4.1_A7_T2.js (previously Failed)
  test/built-ins/RegExp/S15.10.4.1_A5_T8.js [strict mode] (previously Failed)
  test/built-ins/RegExp/S15.10.4.1_A5_T8.js (previously Failed)
  test/built-ins/RegExp/S15.10.4.1_A5_T6.js [strict mode] (previously Failed)
  test/built-ins/RegExp/S15.10.4.1_A5_T6.js (previously Failed)
  test/built-ins/RegExp/S15.10.4.1_A8_T12.js [strict mode] (previously Failed)
  test/built-ins/RegExp/S15.10.4.1_A8_T12.js (previously Failed)
  test/built-ins/RegExp/15.10.4.1-1.js [strict mode] (previously Failed)
  test/built-ins/RegExp/15.10.4.1-1.js (previously Failed)
  test/built-ins/RegExp/S15.10.4.1_A2_T2.js [strict mode] (previously Failed)
  test/built-ins/RegExp/S15.10.4.1_A2_T2.js (previously Failed)
  test/built-ins/RegExp/S15.10.4.1_A5_T4.js [strict mode] (previously Failed)
  test/built-ins/RegExp/S15.10.4.1_A5_T4.js (previously Failed)
  test/built-ins/RegExp/15.10.4.1-3.js [strict mode] (previously Failed)
  test/built-ins/RegExp/15.10.4.1-3.js (previously Failed)
  test/built-ins/RegExp/prototype/Symbol.search/success-get-index-err.js [strict mode] (previously Failed)
  test/built-ins/RegExp/prototype/Symbol.search/success-get-index-err.js (previously Failed)
  test/built-ins/RegExp/prototype/Symbol.search/set-lastindex-init-samevalue.js [strict mode] (previously Failed)
  test/built-ins/RegExp/prototype/Symbol.search/set-lastindex-init-samevalue.js (previously Failed)
  test/built-ins/RegExp/prototype/Symbol.search/set-lastindex-restore-err.js [strict mode] (previously Failed)
  test/built-ins/RegExp/prototype/Symbol.search/set-lastindex-restore-err.js (previously Failed)
  test/built-ins/RegExp/prototype/Symbol.search/set-lastindex-restore-samevalue.js [strict mode] (previously Failed)
  test/built-ins/RegExp/prototype/Symbol.search/set-lastindex-restore-samevalue.js (previously Failed)
  test/built-ins/RegExp/prototype/Symbol.replace/get-unicode-error.js [strict mode] (previously Failed)
  test/built-ins/RegExp/prototype/Symbol.replace/get-unicode-error.js (previously Failed)
  test/built-ins/RegExp/prototype/Symbol.replace/arg-2-coerce-err.js [strict mode] (previously Failed)
  test/built-ins/RegExp/prototype/Symbol.replace/arg-2-coerce-err.js (previously Failed)
  test/built-ins/RegExp/prototype/Symbol.replace/subst-capture-idx-2.js [strict mode] (previously Failed)
  test/built-ins/RegExp/prototype/Symbol.replace/subst-capture-idx-2.js (previously Failed)
  test/built-ins/RegExp/prototype/Symbol.replace/poisoned-stdlib.js [strict mode] (previously Failed)
  test/built-ins/RegExp/prototype/Symbol.replace/poisoned-stdlib.js (previously Failed)
  test/built-ins/RegExp/prototype/Symbol.replace/subst-capture-idx-1.js [strict mode] (previously Failed)
  test/built-ins/RegExp/prototype/Symbol.replace/subst-capture-idx-1.js (previously Failed)
  test/language/expressions/delete/11.4.1-5-a-28-s.js [strict mode] (previously Failed)
  test/language/expressions/delete/11.4.1-5-a-28-s.js (previously Failed)
  test/language/expressions/typeof/built-in-exotic-objects-no-call.js [strict mode] (previously Failed)
  test/language/expressions/typeof/built-in-exotic-objects-no-call.js (previously Failed)
  ```
  </details>
  
  <details><summary><b>Broken tests:</b></summary>
  
  ```
  test/built-ins/RegExp/prototype/Symbol.split/limit-0-bail.js [strict mode] (previously Passed)
  test/built-ins/RegExp/prototype/Symbol.split/limit-0-bail.js (previously Passed)
  ```

HalidOdat commented on boa/src/builtins/regexp/mod.rs:
```diff
@@ -188,114 +185,210 @@ impl RegExp {
             unicode,
-            original_source: regex_body,
-            original_flags: regex_flags,
+            original_source: p,
+            original_flags: f,
```
I think its best to make `original_source` and `original_flags` `JsString`s

raskad replied:
Done in a0d9713.

No changes in the tests expect for `test/built-ins/Map/prototype/keys/returns-iterator.js`. But I think that one is just panicking randomly sometimes.

HalidOdat reviewed (commented):
This looks good :)

Just a small change.

HalidOdat commented on boa/src/builtins/regexp/mod.rs:
```diff
@@ -188,114 +185,210 @@ impl RegExp {
+        let flags = args.get(1).cloned().unwrap_or_else(Value::undefined);
+
+        // 1. Let patternIsRegExp be ? IsRegExp(pattern).
+        let pattern_is_regexp = if let Value::Object(obj) = &pattern {
+            let obj = obj.borrow();
+            obj.is_regexp()
+        } else {
+            false
+        };
```
```suggestion
        let pattern_is_regexp = if let Value::Object(obj) = &pattern {
            let obj = obj.borrow();
            obj.is_regexp()
        	Some(obj.clone())
        } else {
            None
        };
```

So we can avoid calling `.get_field`, we can let it be a `Option<GcObject>`

HalidOdat commented on boa/src/builtins/regexp/mod.rs:
```diff
@@ -188,114 +185,210 @@ impl RegExp {
+        if new_target.is_undefined() {
+            // a. Let newTarget be the active function object.
+            // b. If patternIsRegExp is true and flags is undefined, then
+            if pattern_is_regexp && flags.is_undefined() {
+                // i. Let patternConstructor be ? Get(pattern, "constructor").
+                let pattern_constructor = pattern.get_field("constructor", context)?;
+
+                // ii. If SameValue(newTarget, patternConstructor) is true, return pattern.
+                if Value::same_value(&new_target, &pattern_constructor) {
+                    return Ok(pattern);
                 }
             }
```
```suggestion
            if let Some(pattern) = pattern_is_regexp {
            	if flags.is_undefined() {    	
    	            // i. Let patternConstructor be ? Get(pattern, "constructor").
	                let pattern_constructor = pattern.get("constructor", context)?;

                	// ii. If SameValue(newTarget, patternConstructor) is true, return pattern.
                	if Value::same_value(&new_target, &pattern_constructor) {
                    	return Ok(pattern.into());
                	}
                }
            }
```

HalidOdat commented on boa/src/builtins/regexp/mod.rs:
```diff
@@ -188,114 +185,210 @@ impl RegExp {
-        let mut dot_all = false;
+        // 4. If Type(pattern) is Object and pattern has a [[RegExpMatcher]] internal slot, then
+        // 6. Else,
+        let (p, f) = if pattern_is_regexp {
```
```suggestion
        let (p, f) = if let Some(pattern) = pattern_is_regexp {
```

HalidOdat commented on boa/src/builtins/regexp/mod.rs:
```diff
@@ -1125,19 +1218,24 @@ impl RegExp {
-        let unicode = this.get_field("unicode", context)?.to_boolean();
+        let mut unicode = false;
         if global {
+            // a. Let fullUnicode be ! ToBoolean(? Get(rx, "unicode")).
+            unicode = this.get_field("unicode", context)?.to_boolean();
+
```
```suggestion
            // a. Let fullUnicode be ! ToBoolean(? Get(rx, "unicode")).
            unicode = rx.get("unicode", context)?.to_boolean();

```
We should call `GcObject::get` here. We can should extract the object

```rust
// 1. Let rx be the this value.
// 2. If Type(rx) is not Object, throw a TypeError exception.
let rx = if let Some(rx) = this.as_object() {
	rx
} else {
	return context.throw_type_error(
    	"RegExp.prototype[Symbol.replace] method called on incompatible value",
    );
};
```

HalidOdat commented on boa/src/builtins/regexp/mod.rs:
```diff
@@ -1125,19 +1218,24 @@ impl RegExp {
+            // a. Let fullUnicode be ! ToBoolean(? Get(rx, "unicode")).
+            unicode = this.get_field("unicode", context)?.to_boolean();
+
             // b. Perform ? Set(rx, "lastIndex", +0𝔽, true).
             this.set_field("lastIndex", 0, true, context)?;
```
```suggestion
            // b. Perform ? Set(rx, "lastIndex", +0𝔽, true).
            rx.set("lastIndex", 0, true, context)?;
```
same here

HalidOdat commented on boa/src/builtins/regexp/mod.rs:
```diff
@@ -1366,10 +1464,10 @@ impl RegExp {
         let result = Self::abstract_exec(this, arg_str, context)?;
```
Looking at `RegExpExec` spec, it can only return null or object. We can leverage the rust type system on this and use `Option<GcObject>`

HalidOdat commented on boa/src/builtins/string/mod.rs:
```diff
@@ -865,30 +865,34 @@ impl String {
-        let rx = RegExp::constructor(&Value::from(Object::default()), &[regexp], context)?;
+        let rx = RegExp::create(regexp, Value::undefined(), context)?;
 
         // 5. Return ? Invoke(rx, @@match, « S »).
-        RegExp::r#match(&rx, &[Value::from(arg_str)], context)
+        let obj = rx.as_object().expect("RegExpCreate must return Object");
```
if it only returns object it should return `Result<GcObject>`

raskad commented:
@HalidOdat I rebased this and implemented your suggestions. I refactored a few more things in the RegExp function directly use `GcObject` instead of `Value`. You may want to look at https://github.com/boa-dev/boa/pull/1434/commits/152f081a53210d1b01759d95b41efe1738b4163b to see the refactor.

HalidOdat reviewed (approved):
Looks good to me :)
